sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs

```diff
@@ -0,0 +1,239 @@
+                {
+                    return programState;
+                }
+
+                if (assignmentExpression.Right.IsNullLiteral())
+                {
+                    // The formatter is considered unsafe if the binder is null.
+                    return programState.SetConstraint(formatterSymbolValue, SerializationBinderConstraint.Unsafe);
+                }
+
+                var binderSymbol = semanticModel.GetSymbolInfo(assignmentExpression.Right).Symbol;
+                var binderSymbolValue = programState.GetSymbolValue(binderSymbol);
+                if (binderSymbolValue != null &&
+                    programState.HasConstraint(binderSymbolValue, ObjectConstraint.Null))
+                {
+                    // The formatter is considered unsafe if the binder is null.
+                    return programState.SetConstraint(formatterSymbolValue, SerializationBinderConstraint.Unsafe);
+                }
+
+                var binderType = semanticModel.GetTypeInfo(assignmentExpression.Right).Type;
+                var constraint = IsBinderSafe(binderType)
+                    ? SerializationBinderConstraint.Safe
+                    : SerializationBinderConstraint.Unsafe;
+
+                return programState.SetConstraint(formatterSymbolValue, constraint);
+            }
+
+            private bool HasSafeBinderInitialization(InitializerExpressionSyntax initializer)
+            {
+                var binderAssignment = initializer?.Expressions
+                    .OfType<AssignmentExpressionSyntax>()
+                    .SingleOrDefault(assignment => IsBinderProperty(assignment.Left));
+
+                if (binderAssignment == null ||
+                    binderAssignment.Right.IsNullLiteral())
+                {
+                    return false;
+                }
+
+                var typeSymbol = semanticModel.GetTypeInfo(binderAssignment.Right).Type;
+                return typeSymbol != null && IsBinderSafe(typeSymbol);
+            }
+
+            private bool IsBinderSafe(ITypeSymbol symbol) =>
+                binderValidityMap.GetOrAdd(symbol, typeSymbol =>
+                {
+                    var declaration = GetBindToTypeMethodDeclaration(typeSymbol);
+
+                    // The binder is considered safe by default (e.g. if the declaration cannot be found).
+                    return declaration == null || declaration.ThrowsOrReturnsNull();
+                });
+
+            private static MethodDeclarationSyntax GetBindToTypeMethodDeclaration(ISymbol symbol) =>
+                symbol.DeclaringSyntaxReferences
+                    .SelectMany(GetDescendantNodes)
+                    .OfType<MethodDeclarationSyntax>()
+                    .FirstOrDefault(IsBindToType);
+
+            private static IEnumerable<SyntaxNode> GetDescendantNodes(SyntaxReference syntaxReference) =>
+                syntaxReference.SyntaxTree.GetRoot().FindNode(syntaxReference.Span).DescendantNodes();
```

Comment:
Can't we use
```suggestion
                syntaxReference.GetSyntax().DescendantNodes();
```
instead ?